e2e/nomostest/clean.go

```diff
@@ -126,6 +133,88 @@ func Clean(nt *NT) error {
+	rmObj.SetName(reconcilermanager.ManagerName)
+	if err := nt.KubeClient.Get(rmObj.GetName(), rmObj.GetNamespace(), rmObj); err != nil {
+		return fmt.Errorf("checking reconciler-manager status: %v", err)
+	}
+	if err := testpredicates.StatusEquals(nt.Scheme, kstatus.CurrentStatus)(rmObj); err != nil {
+		return fmt.Errorf("reconciler-manager not healthy: %v", err)
+	}
+	// If reconciler-manager is healthy, uninstall all the unmanaged RootSyncs & RepoSyncs.
+	// If deletion-propagation is enabled, this will also recursively delete all
+	// managed objects, including other managed packages.
+	if err := uninstallUnmanagedPackagesOfType(nt, kinds.RepoSyncV1Beta1()); err != nil {
+		return fmt.Errorf("uninstalling unmanged RepoSyncs: %v", err)
+	}
+	if err := uninstallUnmanagedPackagesOfType(nt, kinds.RootSyncV1Beta1()); err != nil {
+		return fmt.Errorf("uninstalling unmanged RootSyncs: %v", err)
+	}
+	return nil
+}
+
+func uninstallUnmanagedPackagesOfType(nt *NT, gvk schema.GroupVersionKind) error {
+	// Find all packages
+	nt.T.Logf("[CLEANUP] listing %s objects...", gvk.Kind)
+	rsObjList := &unstructured.UnstructuredList{}
+	rsObjList.SetGroupVersionKind(kinds.ListGVKForItemGVK(gvk))
+	if err := nt.KubeClient.List(rsObjList, client.InNamespace(configmanagement.ControllerNamespace)); err != nil {
+		return fmt.Errorf("listing %s: %v", gvk.Kind, err)
+	}
+	// Find the unmanaged packages
+	var rsObjs []*unstructured.Unstructured
+	for _, item := range rsObjList.Items {
+		if core.GetAnnotation(&item, metadata.ResourceManagementKey) == metadata.ResourceManagementEnabled {
+			rsObjs = append(rsObjs, &item)
+		}
+	}
+	// Delete the unmanaged packages in parallel
+	nt.T.Log("[CLEANUP] deleting unmanaged %s objects...", gvk.Kind)
+	tg := taskgroup.New()
+	for _, obj := range rsObjs {
+		nn := client.ObjectKeyFromObject(obj)
+		gvk, err := kinds.Lookup(obj, nt.Scheme)
+		if err != nil {
+			return err
+		}
+		if !obj.GetDeletionTimestamp().IsZero() {
+			nt.T.Logf("[CLEANUP] already terminating %s object %s ...", gvk.Kind, nn)
+		} else {
+			nt.T.Logf("[CLEANUP] deleting %s object %s ...", gvk.Kind, nn)
+			if err := nt.KubeClient.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+				if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+					// skip waiting
+					continue
+				}
+				return fmt.Errorf("unable to delete %s object %s: %w",
+					gvk.Kind, nn, err)
+			}
+		}
+		tg.Go(func() error {
+			nt.T.Logf("[CLEANUP] Waiting for deletion or finalizer failure of %s object %s ...", gvk.Kind, nn)
+			return nt.Watcher.WatchObject(gvk, nn.Name, nn.Namespace,
+				[]testpredicates.Predicate{
+					testpredicates.Or(nt.Logger,
+						testpredicates.ObjectNotFoundPredicate(nt.Scheme),
+						testpredicates.StatusEquals(nt.Scheme, kstatus.FailedStatus),
+						testpredicates.RootSyncHasConditionStatus(v1beta1.RootSyncReconcilerFinalizerFailure, metav1.ConditionTrue),
```

Comment:
Would this predicate apply to RepoSyncs as well?

Reply:
Fixed. Made it generic.